### DISCUSSION TITLE
How to inject custom CA(or anything else) into provider controller image?

### DISCUSSION THREAD

**Discussion by ytsarev** [crossplane/crossplane#2777] [Q&A]
Inspired by working with https://github.com/crossplane-contrib/provider-terraform in a corporate environment where custom Certificate Authority should be part of the image.

My current workaround is a custom provider-terraform-controller image build where CA is getting injected during build time plus associated ControllerConfig with an image override.

It is obviously inconvenient and requires additional maintenance mechanics whenever provider-terraform is updated upstream.

Is there any other more streamlined way to inject the custom CA?

I was trying to find the ability to extend current provider controller deployment with InitContainer which would pull the custom CA during the pod initialization but apparently, it is not available on ControllerConfig abstraction level https://doc.crds.dev/github.com/crossplane/crossplane/pkg.crossplane.io/ControllerConfig/v1alpha1@v1.5.1

Are there any other possibilities to inject custom content into the controller image?

**Comment by hasheddan** [comment]
@ytsarev thanks for raising this question! The "easiest" way to do what you are suggesting today is via `envFrom` with a `Secret` in `ControllerConfig` (https://doc.crds.dev/github.com/crossplane/crossplane/pkg.crossplane.io/ControllerConfig/v1alpha1@v1.5.1#spec-envFrom-secretRef). We could also add the ability to define volume mounts / etc. on `ControllerConfig`, but I see this as a broader issue that I would like us to address in a more extensible manner. I have written up fairly extensive thoughts on the matter in https://github.com/crossplane/crossplane/issues/2671, but the summary is that I believe users are going to want to be able to configure provider controllers in a more robust manner (exemplified by what you want to do here), including running provider controllers externally to the cluster. This would be offered in the form of an interface (`Provider Runtime Interface (PRI)`), of which the current in-cluster `Deployment` strategy would be one implementation. Each implementation could offer its own configuration options, meaning that if you didn't like the way the default system allowed you to run controllers, you could swap it out or create your own.

**Reply by ytsarev** [reply]
@hasheddan thanks a lot for the detailed response! I am afraid environment variable is not enough here as in case of custom CA file we need to:
1. Mount/put the file under system SSL directory like `/usr/local/share/ca-certificates/`
2. Run `update-ca-certificates`

That's why I think InitContainer might generally solve it.

**Reply by marshmallory** [reply]
We also need this functionality for local development behind Zscaler. 

Zscaler in particular also will not work off an environment variable.  Instead, it has to be injected.

Going to use Yury's suggestion around building it ourselves but this feature would be awesome for enterprises!

**Comment by luebken** [comment]
@muvaf Is this something that can be solved / made easier with your Webhooks foundation work?

**Comment by luebken** [comment]
See also https://github.com/crossplane/crossplane/issues/2535

**Comment by ytsarev** [comment]
Maybe https://github.com/microcumulus/ca-injector can help

**Comment by marshmallory** [comment]
Bump.  Working with Stephen at the moment, but there's no way to override this at all.  Have done a bunch of troubleshooting on our end to try to inject the CA bundle with no luck.

We're able to get it to communicate well with AWS up until the `terraform plan`.  After that, we get an x509 error.

**Comment by ytsarev** [comment]
Solved with the new VolumeMounts extension of ControllerConfig. 

See the guide in the gist below
https://gist.github.com/stevendborrelli/582fb81b844a76edd6eb3ab00d5b45b6

Thanks @stevendborrelli !

**Comment by jbw976** [comment]
Looks like good content for the crossplane docs - @stevendborrelli would you be willing to open a PR for that? 💪

**Reply by stevendborrelli** [reply]
Yes, I'll create a PR.

**Reply by haarchri** [reply]
https://github.com/crossplane/docs/pull/845
